### PR TITLE
Fix prodlike workflows

### LINKED ISSUE #35966: Run3 production-like workflow contains duplicated NANO step since 12_1_0_pre5
We noticed that in 12_1_0_pre5, when running

```
runTheMatrix.py -w upgrade -l 11834.21 --dryRun
```

the output contains a new "step6", which is a duplicate of the previous NANO step, without some PU flags. Is this expected?

```
Preparing to run 11834.21 TTbar_14TeV+2021PU_ProdLike+TTbar_14TeV_TuneCP5_GenSim+DigiPU+RecoPU+MiniAODPU+NanoPU+Nano

# in: /afs/cern.ch/work/j/jpata/CMSSW_12_1_0_pre5 dryRun for 'cd 11834.21_TTbar_14TeV+2021PU_ProdLike+TTbar_14TeV_TuneCP5_GenSim+DigiPU+RecoPU+MiniAODPU+NanoPU+Nano
 cmsDriver.py TTbar_14TeV_TuneCP5_cfi  -s GEN,SIM -n 10 --conditions auto:phase1_2021_realistic --beamspot Run3RoundOptics25ns13TeVLowSigmaZ --datatier GEN-SIM --eventcontent FEVTDEBUG --geometry DB:Extended --era Run3 --relval 9000,100 --fileout file:step1.root  > step1_TTbar_14TeV+2021PU_ProdLike+TTbar_14TeV_TuneCP5_GenSim+DigiPU+RecoPU+MiniAODPU+NanoPU+Nano.log  2>&1
 

# in: /afs/cern.ch/work/j/jpata/CMSSW_12_1_0_pre5 dryRun for 'cd 11834.21_TTbar_14TeV+2021PU_ProdLike+TTbar_14TeV_TuneCP5_GenSim+DigiPU+RecoPU+MiniAODPU+NanoPU+Nano
 cmsDriver.py step2  -s DIGI,L1,DIGI2RAW,HLT:@relval2021 --conditions auto:phase1_2021_realistic --datatier GEN-SIM-RAW -n 10 --eventcontent RAWSIM --geometry DB:Extended --era Run3 --pileup Run3_Flat55To75_PoissonOOTPU --pileup_input das:/RelValMinBias_14TeV/CMSSW_12_0_0_pre4-120X_mcRun3_2021_realistic_v2-v1/GEN-SIM --filein  file:step1.root  --fileout file:step2.root  > step2_TTbar_14TeV+2021PU_ProdLike+TTbar_14TeV_TuneCP5_GenSim+DigiPU+RecoPU+MiniAODPU+NanoPU+Nano.log  2>&1
 

# in: /afs/cern.ch/work/j/jpata/CMSSW_12_1_0_pre5 dryRun for 'cd 11834.21_TTbar_14TeV+2021PU_ProdLike+TTbar_14TeV_TuneCP5_GenSim+DigiPU+RecoPU+MiniAODPU+NanoPU+Nano
 cmsDriver.py step3  -s RAW2DIGI,L1Reco,RECO,RECOSIM --conditions auto:phase1_2021_realistic --datatier AODSIM -n 10 --eventcontent AODSIM --geometry DB:Extended --era Run3 --pileup Run3_Flat55To75_PoissonOOTPU --pileup_input das:/RelValMinBias_14TeV/CMSSW_12_0_0_pre4-120X_mcRun3_2021_realistic_v2-v1/GEN-SIM --filein  file:step2.root  --fileout file:step3.root  > step3_TTbar_14TeV+2021PU_ProdLike+TTbar_14TeV_TuneCP5_GenSim+DigiPU+RecoPU+MiniAODPU+NanoPU+Nano.log  2>&1
 

# in: /afs/cern.ch/work/j/jpata/CMSSW_12_1_0_pre5 dryRun for 'cd 11834.21_TTbar_14TeV+2021PU_ProdLike+TTbar_14TeV_TuneCP5_GenSim+DigiPU+RecoPU+MiniAODPU+NanoPU+Nano
 cmsDriver.py step4  -s PAT --conditions auto:phase1_2021_realistic --datatier MINIAODSIM -n 10 --eventcontent MINIAODSIM --geometry DB:Extended --era Run3 --pileup Run3_Flat55To75_PoissonOOTPU --pileup_input das:/RelValMinBias_14TeV/CMSSW_12_0_0_pre4-120X_mcRun3_2021_realistic_v2-v1/GEN-SIM --filein  file:step3.root  --fileout file:step4.root  > step4_TTbar_14TeV+2021PU_ProdLike+TTbar_14TeV_TuneCP5_GenSim+DigiPU+RecoPU+MiniAODPU+NanoPU+Nano.log  2>&1
 

# in: /afs/cern.ch/work/j/jpata/CMSSW_12_1_0_pre5 dryRun for 'cd 11834.21_TTbar_14TeV+2021PU_ProdLike+TTbar_14TeV_TuneCP5_GenSim+DigiPU+RecoPU+MiniAODPU+NanoPU+Nano
 cmsDriver.py step5  -s NANO --conditions auto:phase1_2021_realistic --datatier NANOAODSIM -n 10 --eventcontent NANOEDMAODSIM --filein file:step4.root --geometry DB:Extended --era Run3 --pileup Run3_Flat55To75_PoissonOOTPU --pileup_input das:/RelValMinBias_14TeV/CMSSW_12_0_0_pre4-120X_mcRun3_2021_realistic_v2-v1/GEN-SIM --fileout file:step5.root  > step5_TTbar_14TeV+2021PU_ProdLike+TTbar_14TeV_TuneCP5_GenSim+DigiPU+RecoPU+MiniAODPU+NanoPU+Nano.log  2>&1
 

# in: /afs/cern.ch/work/j/jpata/CMSSW_12_1_0_pre5 dryRun for 'cd 11834.21_TTbar_14TeV+2021PU_ProdLike+TTbar_14TeV_TuneCP5_GenSim+DigiPU+RecoPU+MiniAODPU+NanoPU+Nano
 cmsDriver.py step6  -s NANO --conditions auto:phase1_2021_realistic --datatier NANOAODSIM -n 10 --eventcontent NANOEDMAODSIM --filein file:step4.root --geometry DB:Extended --era Run3 --fileout file:step6.root  > step6_TTbar_14TeV+2021PU_ProdLike+TTbar_14TeV_TuneCP5_GenSim+DigiPU+RecoPU+MiniAODPU+NanoPU+Nano.log  2>&1
 
11834.21_TTbar_14TeV+2021PU_ProdLike+TTbar_14TeV_TuneCP5_GenSim+DigiPU+RecoPU+MiniAODPU+NanoPU+Nano Step0-PASSED Step1-PASSED Step2-PASSED Step3-PASSED Step4-PASSED Step5-PASSED  - time date Wed Nov  3 14:40:52 2021-date Wed Nov  3 14:40:52 2021; exit: 0 0 0 0 0 0
1 1 1 1 1 1 tests passed, 0 0 0 0 0 0 failed
```

cc @cms-sw/xpog-l2 @cms-sw/pdmv-l2

### PR DESCRIPTION
**Pull request by kskovpen**
#### PR description:

Hopefully, fixes #35966 

#### PR validation:

Tested on standard and prodlike wfs

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Not a backport

### CODE REVIEW

**Comment by kskovpen** [comment]
assign reconstruction

**Comment by kskovpen** [comment]
please test

**Comment by srimanob** [comment]
@cmsbuild please test with https://github.com/cms-sw/cmssw/pull/36021

**Comment by kskovpen** [comment]
please test workflow 11834.21 with #36021

**Comment by kskovpen** [comment]
Not sure what failed in Jenkins as it seems to show OK status

**Comment by kskovpen** [comment]
Looks like this specific workflow is failing because Jenkins runs relval workflows in standard setup. If anyone knows how to trigger the upgrade ones via Jenkins, please let me know :)

**Comment by srimanob** [comment]
@kskovpen 
Not possible, you need to define it first in 
https://github.com/cms-sw/cmssw/blob/master/Configuration/PyReleaseValidation/python/relval_2017.py

Maybe update this file also, then it should be able to trigger the test.

**Comment by kpedro88** [comment]
@kskovpen the bot now supports upgrade workflows, see e.g. https://github.com/cms-sw/cmssw/pull/35104#issuecomment-910563388

**Comment by kskovpen** [comment]
> @kskovpen the bot now supports upgrade workflows, see e.g. [#35104 (comment)](https://github.com/cms-sw/cmssw/pull/35104#issuecomment-910563388)

oh wow that's great

**Comment by kskovpen** [comment]
test parameters:

- workflow=11834.21
- relvals_opt= -w upgrade

**Comment by kskovpen** [comment]
-1

File: Configuration/PyReleaseValidation/python/relval_upgrade.py [hunk 1]
```diff
@@ -47,7 +47,8 @@ def makeStepName(key,frag,step,suffix):
                         if 'ProdLike' in specialType:
                             if 'Reco' in step: # handles both Reco and RecoGlobal
                                 stepList[specialType].append(stepMaker(key,frag[:-4],step.replace('RecoGlobal','MiniAOD').replace('Reco','MiniAOD'),specialWF.suffix))
-                                stepList[specialType].append(stepMaker(key,frag[:-4],step.replace('RecoGlobal','Nano').replace('Reco','Nano'),specialWF.suffix))
+                                if 'PU' not in step:
```
Comment (kpedro88) [inline Configuration/PyReleaseValidation/python/relval_upgrade.py:50]: just for my own understanding, can you explain why Reco+Nano are combined in noPU workflows, but separate in PU workflows?

**Comment by kskovpen** [comment]
test parameters:

workflow=11834.21
relvals_opt= -w upgrade
pull_request=36021

**Comment by kskovpen** [comment]
please test

**Comment by kpedro88** [comment]
Looking into this further, I think the actual problem is that the addition of Nano in #35412 just wasn't implemented consistently. I will try to come up with the right way to do it, but it would be useful if @kskovpen @cms-sw/pdmv-l2 @cms-sw/xpog-l2 could indicate what they *want* to happen (which years, which workflows, etc.).

**Comment by kskovpen** [comment]
> Looking into this further, I think the actual problem is that the addition of Nano in #35412 just wasn't implemented consistently. I will try to come up with the right way to do it, but it would be useful if @kskovpen @cms-sw/pdmv-l2 @cms-sw/xpog-l2 could indicate what they _want_ to happen (which years, which workflows, etc.).

Thanks @kpedro88. The general idea was to add NanoDQM to relval wfs (2021) and run it together with MiniAOD DQM as one step, while doing a similar thing for Reco and Nano wfs. Within the same PR the Nano step was added to the production wfs. I also need to think if there is a better way to rearrange a bit the general implementation.

**Comment by kpedro88** [comment]
I *think* this does it: https://github.com/cms-sw/cmssw/compare/master...kpedro88:Phase2-WF75v2. Please take a look. (This branch is on top of 12_1_0_pre5.)

**Comment by kskovpen** [comment]
> I _think_ this does it: [master...kpedro88:Phase2-WF75v2](https://github.com/cms-sw/cmssw/compare/master...kpedro88:Phase2-WF75v2). Please take a look. (This branch is on top of 12_1_0_pre5.)

Thanks, @kpedro88. This shall indeed be a bit more consistent across various workflows. I am wondering if having both RecoNano and MiniAOD+Nano within the same workflow is really optimal, since we run MiniAOD and Nano steps twice. Let's see anyway how validation goes.

**Comment by kskovpen** [comment]
please test

**Comment by kskovpen** [comment]
Let the saga continue.

**Comment by kskovpen** [comment]
please test please test

**Comment by kskovpen** [comment]
It looks like there is a deprecated option in the runTheMatrix tests in Jenkins, i.e. '-s', which makes the tests fail - can also be reproduced locally. Removing this option makes the tests run fine. Any ideas anyone?

runTheMatrix.py -l 11834.21 -w upgrade (success)
runTheMatrix.py -s -l 11834.21 -w upgrade (fails)

**Comment by slava77** [comment]
> It looks like there is a deprecated option in the runTheMatrix tests in Jenkins, i.e. '-s', which makes the tests fail - can also be reproduced locally. Removing this option makes the tests run fine. Any ideas anyone?
> 
> runTheMatrix.py -l 11834.21 -w upgrade (success) runTheMatrix.py -s -l 11834.21 -w upgrade (fails)

is it from here
https://github.com/cms-sw/cms-bot/blob/a1194080497a13d952360c64a2e60643f90108fc/run-tests-release#L38
?

**Comment by slava77** [comment]
> > It looks like there is a deprecated option in the runTheMatrix tests in Jenkins, i.e. '-s', which makes the tests fail - can also be reproduced locally. Removing this option makes the tests run fine. Any ideas anyone?
> > runTheMatrix.py -l 11834.21 -w upgrade (success) runTheMatrix.py -s -l 11834.21 -w upgrade (fails)
> 
> is it from here https://github.com/cms-sw/cms-bot/blob/a1194080497a13d952360c64a2e60643f90108fc/run-tests-release#L38 ?

actually, no, this comes from  the jenkins side https://cmssdt.cern.ch/jenkins/job/ib-run-pr-relvals/12635/parameters/

the short matrix should still run, which is what `-s` is for. Does `-l limited,11834.21 -w upgrade` work?

**Comment by kskovpen** [comment]
> > > It looks like there is a deprecated option in the runTheMatrix tests in Jenkins, i.e. '-s', which makes the tests fail - can also be reproduced locally. Removing this option makes the tests run fine. Any ideas anyone?
> > > runTheMatrix.py -l 11834.21 -w upgrade (success) runTheMatrix.py -s -l 11834.21 -w upgrade (fails)
> > 
> > 
> > is it from here https://github.com/cms-sw/cms-bot/blob/a1194080497a13d952360c64a2e60643f90108fc/run-tests-release#L38 ?
> 
> actually, no, this comes from the jenkins side https://cmssdt.cern.ch/jenkins/job/ib-run-pr-relvals/12635/parameters/
> 
> the short matrix should still run, which is what `-s` is for. Does `-l limited,11834.21 -w upgrade` work?

Nope, same issue. Also fails in older releases.

**Comment by kskovpen** [comment]
please test

**Comment by kskovpen** [comment]
I am wondering if there is anything to do with a newline character that got into Jenkins tests after 'runTheMatrix.py -j 16 -s -l ', as can be seen in https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6d03ff/20560/runTheMatrix-results/matrixTests.log. For example, one can compare it to https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fe8735/20522/runTheMatrix-results/matrixTests.log

**Comment by kpedro88** [comment]
@kskovpen:
>  I am wondering if having both RecoNano and MiniAOD+Nano within the same workflow is really optimal, since we run MiniAOD and Nano steps twice.

I don't understand this comment. There should be relval-style workflows that have RecoNano (everything in one step) and production-style workflows that have Reco, Mini, Nano in separate steps. No step should be duplicated within a workflow.

**Comment by kskovpen** [comment]
> @kskovpen:
> 
> > I am wondering if having both RecoNano and MiniAOD+Nano within the same workflow is really optimal, since we run MiniAOD and Nano steps twice.
> 
> I don't understand this comment. There should be relval-style workflows that have RecoNano (everything in one step) and production-style workflows that have Reco, Mini, Nano in separate steps. No step should be duplicated within a workflow.

What I was referring to is that there were duplicated steps within a single production wf. Let's see if things get fixed in the next update.

**Comment by kskovpen** [comment]
please test

**Comment by kpedro88** [comment]
@kskovpen this PR is still not consistent with my branch. e.g. having a Reco step that doesn't have Reco in the name ("RecNan") can cause problems for custom workflows that check for this. Please use the consistent implementation that I provided instead of making haphazard changes.

**Comment by kpedro88** [comment]
some changes from https://github.com/cms-sw/cmssw/compare/master...kpedro88:Phase2-WF75v2 are still missing here. what is the problem?

**Comment by kpedro88** [comment]
superseded by #36167, please close this.

**Pull request closed without merging**